File: bin/git-prompt.go
///usr/bin/true; exec /usr/bin/env go run $0 $@

package main

import (
	"os"
	"os/exec"
)

func main() {
	var prompt []byte

	commit, err_not_a_repo := exec.Command("git", "rev-parse", "HEAD").Output()

	if err_not_a_repo != nil {
		os.Exit(0)
	}

	branch, _ := exec.Command("git", "branch", "--show-current").Output()

	if len(branch) > 0 {
		var fmt_branch = branch[:len(branch)-1] // trim newline
		prompt = append(prompt, fmt_branch...)
	} else {
		var fmt_commit = commit[:6]
		prompt = append(prompt, fmt_commit...)
	}

	status, _ := exec.Command("git", "status", "--porcelain").Output()

	if len(status) > 0 {
		prompt = append(prompt, '*')
	}

	os.Stdout.Write(prompt)
}

File: bin/git-prompt.jl
#!/bin/sh
#=
exec j! $0 $@
=#

"Run a command, discarding stderr, returning stdout as a string"
function execute(cmd::Cmd)
  out = Pipe()
  process = run(pipeline(ignorestatus(cmd), stdout=out, stderr=devnull))
  close(out.in)
  stdout = String(read(out))
end

commit = execute(`git rev-parse HEAD`)

if length(commit) == 0
  exit(0)
end

branch = execute(`git branch --show-current`)

if length(branch) > 0
  prompt = branch[1:length(branch)-1] # trim newline
else
  prompt = commit[1:6]
end

status = execute(`git status --porcelain`)

if length(status) > 0
  prompt *= "*"
end

write(stdout, prompt)
